372 ffd bsplines

acasamitjana opened the pull request:
# Description

Add B-Splines as interpolation method for DDF model

Fixes #372 

## Type of change

What types of changes does your code introduce to DeepReg? _Put an `x` in the boxes that
apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Documentation Update (fix or improvement on the documentation)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Other (if none of the other choices apply)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR.
If you're unsure about any of them, don't hesitate to ask. We're here to help! This is
simply a reminder of what we are going to look for before merging your code._

- [X] I have
      [installed pre-commit](https://deepreg.readthedocs.io/en/latest/contributing/setup.html)
      using `pre-commit install` and formatted all changed files. If you are not
      certain, run `pre-commit run --all-files`.
- [X] My commits' message styles matches
      [our requested structure](https://deepreg.readthedocs.io/en/latest/contributing/commit.html),
      e.g. `Issue #<issue number>: detailed message`.
- [ ] I have updated the
      [change log file](https://github.com/DeepRegNet/DeepReg/blob/main/CHANGELOG.md)
      regarding my changes.
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

acasamitjana commented:
@YipengHu @ebonmati @RemiDelaunay hey, can you take a look at this PR? I include BSplines interpolation in the pipeline. I added a small change in the config file dictionary so probably some test may fail.

mathpluscode commented on deepreg/model/layer.py:
```diff
@@ -665,3 +666,123 @@ def call(self, inputs, training=None, **kwargs):
+
+    def __init__(self, cp_spacing: (int, tuple), **kwargs):
+
+        super().__init__(**kwargs)
```
Here maybe it's better to put `super(BSplines3DTransform, self)` instead of `super()` ?

mathpluscode replied:
lol this is not addressed?

mathpluscode commented on deepreg/model/layer.py:
```diff
@@ -665,3 +666,123 @@ def call(self, inputs, training=None, **kwargs):
+        :return:
+        """
+
+        super().build(input_shape)
```
same comment as the super above

mathpluscode commented on deepreg/model/layer.py:
```diff
@@ -665,3 +666,123 @@ def call(self, inputs, training=None, **kwargs):
+            self.cp_spacing = cp_spacing
+        self.filters = []
+
+    def build(self, input_shape):
```
if you don't mind, put `input_shape:tuple` in the input arg, so that IDE can help people to identify potential bugs

mathpluscode commented on deepreg/model/layer.py:
```diff
@@ -665,3 +666,123 @@ def call(self, inputs, training=None, **kwargs):
+     It assumes a full sized image from which: (1) it compute the contol points values by downsampling the initial
+     image (2) performs the interpolation and (3) crops the image around the valid values.
+
+    :param cp_spacing: _int_ or tuple of three _ints_ specifying the spacing (in pixels) in each dimension.
```
any reason for `_int_` instead of `int` here in the docstring?

mathpluscode commented on deepreg/model/layer.py:
```diff
@@ -665,3 +666,123 @@ def call(self, inputs, training=None, **kwargs):
+            dtype=np.float32,
+        )
+
+        for u in range(self.cp_spacing[0]):
```
these for loops scared me a bit... do we have a better way to do so?

I think this example might inspire you a bit?

https://stackoverflow.com/questions/798854/all-combinations-of-a-list-of-lists

acasamitjana replied:
I simplified the loops, take a look at the updated file

YipengHu replied:
i don't think we can eyeball this to say if it is correctly implemented. Is it possible to use, say sitk.BSplineTransform in the test?

mathpluscode commented on deepreg/model/network/ddf_dvf.py:
```diff
@@ -17,7 +17,7 @@ def ddf_dvf_forward(
     moving_image_size: tuple,
     fixed_image_size: tuple,
-    output_dvf: bool,
+    output_model: (str, None),
```
warning: potential conflicts with #537, we will sort it out later, no worries

acasamitjana replied:
ok, I leave it like this

mathpluscode commented on test/unit/test_layer.py:
```diff
@@ -348,3 +348,34 @@ def test_localNetUpSampleResnetBlock():
+
+
+def test_BSplines3DTransform():
+    """
```
Here you are testing multiple different cases, please split tests into diff funcs inside a big `TestBSplines3DTransform`

check the doc/example here

https://deepreg.readthedocs.io/en/latest/contributing/test.html

mathpluscode reviewed (changes requested):
- Simplification of the for loops
- Refactor unit-tests
- some other small comments/discussions

mathpluscode commented:
@acasamitjana I've left my comments ;)

acasamitjana commented:
> @acasamitjana I've left my comments ;)

Thank you! I have addressed them. @YipengHu you may want to take a look into that as you suggested the bsplines functionality in the first place

mathpluscode commented:
> > @acasamitjana I've left my comments ;)
> 
> Thank you! I have addressed them. @YipengHu you may want to take a look into that as you suggested the bsplines functionality in the first place

@acasamitjana I think you didn't push :D

acasamitjana commented:
> > > @acasamitjana I've left my comments ;)
> > 
> > 
> > Thank you! I have addressed them. @YipengHu you may want to take a look into that as you suggested the bsplines functionality in the first place
> 
> @acasamitjana I think you didn't push :D

ouch sorry, I didn't realize I mispelled the username when pushing LOL. It's done now. I'll look for a comparison with other libraries as Yipeng suggested

acasamitjana commented:
Following Yipeng's suggestion I am comparing against SimpleITK library and getting different results. However, results are consistent with the formula in  https://ieeexplore.ieee.org/stamp/stamp.jsp?tp=&arnumber=796284
I dont know how ITK implemented so I'm not sure what are we comparing with.

I have implemented a small code snippet for a 2D case under test/output that you can check. The file is not passing all pre-commit due to using simpleITK, a library not included in the requirements.

mathpluscode commented on deepreg/model/layer.py:
```diff
@@ -665,3 +668,204 @@ def call(self, inputs, training=None, **kwargs):
+        v_arange = np.arange(0, 1, 1 / self.cp_spacing[1])
+        w_arange = np.arange(0, 1, 1 / self.cp_spacing[2])
+
+        filter_idx = [[0, 1, 2, 3] for _ in range(3)]
```
it can be simpler as `[0, 1, 2, 3] * 3` lol but as you wish, it's super minor suggestion

mathpluscode commented on deepreg/model/layer.py:
```diff
@@ -665,3 +668,204 @@ def call(self, inputs, training=None, **kwargs):
+            3: lambda u: np.float32(u ** 3 / 6),
+        }
+
+        filters = np.zeros(
```
I'm not sure I can understand this by reading the code, is there any docs as ref? I'd be happy to read it and double check the code. For now it's kind of like LGTM lol

btw do you think it will be possible to isolate the code from `filters = np.zeros` to the line before `self.filter = tf.convert_to_tensor(filters)`? so that we can test the values not just the shapes.

mathpluscode replied:
or maybe what we can do for testing the value, is to compare the results of current code with the results of your previous code in test.

so in test, we call this more efficient code, and compare with the easier to understand code, if their results are the same, it validates your efficient code.

mathpluscode replied:
https://ieeexplore.ieee.org/stamp/stamp.jsp?tp=&arnumber=796284 is this the reference?

acasamitjana replied:
Yes, I compared it with the commented code, which is more "literal" but less efficient and it gives the same results.

Regarding the implementation yes, the reference is correct
![comparison](https://user-images.githubusercontent.com/19728301/101179786-fc1a6480-364a-11eb-9d24-42118a3439ac.png)

mathpluscode commented:
> Following Yipeng's suggestion I am comparing against SimpleITK library and getting different results. However, results are consistent with the formula in https://ieeexplore.ieee.org/stamp/stamp.jsp?tp=&arnumber=796284
> I dont know how ITK implemented so I'm not sure what are we comparing with.
> 
> I have implemented a small code snippet for a 2D case under test/output that you can check. The file is not passing pre-commit due to using simpleITK and I think this blocks from pushing, is that right? I can copy the file here if you like

Sorry for the delay, I added some more comments, more is about how can we test that for loop code more rigorously.

mathpluscode commented on deepreg/model/layer.py:
```diff
@@ -665,3 +668,204 @@ def call(self, inputs, training=None, **kwargs):
+            dtype=np.float32,
+        )
+
+        # for u in range(self.cp_spacing[0]):
```
later in the end, before PR, we should delete these commented out code

acasamitjana replied:
Yes, I'll do that

acasamitjana commented:
Hi, I've changed certain things according to an update in the main branch. Commits are coming:
   - Now the method won't be a dictionary, just the name "dvf", "ddf", ... (just as before, hence backward compatible)
   - Instead, the control point spacing would be specified in the config file under the backbone field and computed in the backbone network (I added the instructions at the end)
   - I decoupled the Resize and the interpolation using B-splines. The Resize function may be interesting for other purposes as well.
   - I implemented the tests for the gaussian kernel computation, for the resize and B-splines layers.

Ideally, the best option would be to add another step between build_backbone() and build_ddf_dvf() called build_control_points() in order to be independent of the backbone. Let's discuss this before I implement the additional tests in test_backbone.

mathpluscode commented on deepreg/model/layer.py:
```diff
@@ -665,3 +668,228 @@ def call(self, inputs, training=None, **kwargs):
+            padding="VALID",
+        )
+
+    # def gaussian_filter_3d(self, kernel_sigma):
```
maybe we should remove these comments :)

mathpluscode commented:
@acasamitjana sorry for the delay, the code looks ok for me, if you can remove those commented codes :)

acasamitjana commented:
I forgot to mention the important thing:
- I compared to Simple ITK and the results were solid. I can't push this and I think it is because SimpleITK is not in the requirements. 
- It took me a while to spot the initial difference I was having between both. FYI, SImpleITK considers the center of each pixel at 0.5, 1.5, etc... and not 0.0, 1.0, 2.0. I left the code sampling at 0.5...., you can see it on lines 721-7219 from layers.py

acasamitjana commented:
> @acasamitjana sorry for the delay, the code looks ok for me, if you can remove those commented codes :)

Thanks I'll do. @YipengHu can you take a look at that? No rush

mathpluscode commented:
@acasamitjana I just found that the changes in this PR are not really backbone-dependent?

It would be helpful if we use DDF or DVF, where we can predict low resolution DDF/DVF, then resize. But it does not depend on which backbone we use, local net or u-net?

YipengHu commented:
> @acasamitjana @YipengHu I just found that the changes in this PR are not really backbone-dependent?
> 
> It would be helpful if we use DDF or DVF, where we can predict low resolution DDF/DVF, then resize. But it does not depend on which backbone we use, local net or u-net?

you are correct. the choice between displacement field and velocity field should not be dependent on the network architecture.
